quality_of_service_demo/rclcpp/src/interactive_publisher.cpp

```diff
@@ -157,15 +157,20 @@ int main(int argc, char * argv[])
 
   PublisherCommandHandler cmd_handler(exec, talker);
 
-  talker->initialize();
-  talker->print_qos();
-
-  cmd_handler.start();
-  exec.add_node(talker);
-  exec.spin();
-  exec.remove_node(talker);
-  cmd_handler.stop();
+  try
+  {
+    talker->initialize();
+    talker->print_qos();
 
+    cmd_handler.start();
+    exec.add_node(talker);
+    exec.spin();
+  }
+  catch (int n)
```

Comment:
Sorry, I don't understand the fix.
Where this exception was thrown?

Reply:
I'm trying to use a generic exception, but I guess `int n` is not the right one. I have updated it to `const std::exception&`.

Reply:
You can just use:
```cpp
} catch (...) {
```

Though catching particularly `std::exception` allows you to print a better error message (exceptions following the `std::exception` interface have a [what](https://en.cppreference.com/w/cpp/error/exception/what) method).
Sometimes, when people want to catch everything and provide as much information as possible they do something like [this](https://github.com/ros2/rclcpp/blob/0c66d0c725ad1f977e2b907bc0619d6ceb6de238/rclcpp/src/rclcpp/context.cpp#L47-L51).

----

I still don't see how this solves #407, can you explain further?

Reply:
I'm trying to mimic the behaviour of `try/except` block like in this python [example](https://github.com/ros2/demos/blob/master/demo_nodes_py/demo_nodes_py/topics/listener.py#L35). I think by treating `ctrl-c` as a keyboard exception, the process is able to run the `exec.remove_node` and `cmd_handler.stop` immediately instead of waiting for the separate thread. After testing it on ubuntu 18.04, it seems to work. Thus I made the PR. There could be a better way to solve this problem. This is just my attempt to pitch something that worked for me.

Reply:
cpp doesn't throw an exception when treating ctrl-c, does something Python specific.
If you print something in the catch statement, you probably won't see anything.

The only way of catching a signal in cpp is by installing a signal handler, and as `rclcpp` already installs one, you will have to do chaining. See https://github.com/ros2/rclcpp/blob/master/rclcpp/src/rclcpp/signal_handler.cpp.

Reply:
@ivanpauno is correct. This patch as-is won't do.

Reply:
I have backlogged this PR cuz I don't have time to work on it in recent weeks, and the process can be killed if use `ctrl-c` and `x` as discussed in the original issue. I'll come back to it at some point...